pilot/pkg/serviceregistry/kube/controller/ambientindex_external.go

```diff
@@ -37,24 +40,175 @@ import (
 	"istio.io/istio/pkg/workloadapi"
 )
 
-func (c *Controller) getWorkloadEntriesInPolicy(ns string, sel map[string]string) []*apiv1alpha3.WorkloadEntry {
-	if ns == c.meshWatcher.Mesh().GetRootNamespace() {
-		ns = metav1.NamespaceAll
+func (a *AmbientIndexImpl) HandleServiceEntry(svc *model.Service, event model.Event, c *Controller) {
+	if svc.Attributes.ServiceEntry == nil {
+		// event for e.g. kube svc; ignore
+		return
 	}
 
-	allWorkloadEntries := c.getControllerWorkloadEntries(ns)
-	var workloadEntries []*apiv1alpha3.WorkloadEntry
-	for _, w := range allWorkloadEntries {
-		if labels.Instance(sel).SubsetOf(w.Spec.Labels) {
-			workloadEntries = append(workloadEntries, w)
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// We will accrue updates as we update our internal state
+	updates := sets.New[model.ConfigKey]()
+
+	if event != model.EventAdd {
+		a.cleanupOldWorkloadEntriesInlinedOnServiceEntry(svc, updates, c)
+	}
+
+	serviceEntryNamespacedName := types.NamespacedName{
+		Name:      svc.Attributes.ServiceEntryName,
+		Namespace: svc.Attributes.ServiceEntryNamespace,
+	}
+
+	// Update indexes
+	if event == model.EventDelete {
+		// servicesMap is used when cleaning up old WEs inlined on a SE (i.e., `ServiceEntry.endpoints`)
+		// so we must delete this after we clean up the old WEs. That way we don't miss any auto-allocated
+		// VIPs during cleanup on the idx.byWorkloadEntry[networkAddr] map
+		delete(a.servicesMap, serviceEntryNamespacedName)
+	} else {
+		// servicesMap is used when constructing workloads so it must be up to date
+		a.servicesMap[serviceEntryNamespacedName] = svc
+	}
+
+	sel := klabels.Set(svc.Attributes.ServiceEntry.WorkloadSelector.GetLabels()).AsSelectorPreValidated()
```

Comment:
nit: us ValidatedSetSelector instead, its (much) faster

Reply:
addressed in https://github.com/istio/istio/pull/45621/commits/b6327754170bdb16e1d805bde7956f01664089b3